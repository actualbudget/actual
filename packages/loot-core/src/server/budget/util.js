import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return amounts.reduce((total, amount) => {
    return total + number(amount);
  }, 0);
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}

// Note that we don't restrict values to `Number.MIN_SAFE_INTEGER <= value <= Number.MAX_SAFE_INTEGER`
// where `Number.MAX_SAFE_INTEGER == 2^53 - 1` but a smaller range over `-(2^43-1) <= value <= 2^43 - 1`.
// This ensure that the number is accurate not just for the integer component but for 3 decimal places also.
//
// This gives us the guarantee that can use `safeNumber` on number whether they are unscaled user inputs
// or they have been converted to integers (using `amountToInteger`).

const MAX_SAFE_NUMBER = 2 ** 43 - 1;
const MIN_SAFE_NUMBER = -MAX_SAFE_NUMBER;

export function safeNumber(value) {
  value = number(value);
  if (value > MAX_SAFE_NUMBER || value < MIN_SAFE_NUMBER) {
    throw new Error(
      "Can't safely perform arithmetic with number: " + JSON.stringify(value)
    );
  }
  return value;
}
