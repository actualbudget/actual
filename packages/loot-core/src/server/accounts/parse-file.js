import csv2json from 'csv-parse/lib/sync';

import * as fs from '../../platform/server/fs';
import { dayFromDate } from '../../shared/months';
import { looselyParseAmount } from '../../shared/util';

import qif2json from './qif2json';

export function parseFile(filepath, options) {
  let errors = [];
  let m = filepath.match(/\.[^.]*$/);

  if (m) {
    let ext = m[0];

    switch (ext.toLowerCase()) {
      case '.qif':
        return parseQIF(filepath);
      case '.csv':
        return parseCSV(filepath, options);
      case '.ofx':
      case '.qfx':
        return parseOFX(filepath);
      default:
    }
  }

  errors.push({
    message: 'Invalid file type',
    internal: '',
  });
  return { errors };
}

async function parseCSV(filepath, options = {}) {
  let errors = [];
  let contents = await fs.readFile(filepath);

  let data;
  try {
    data = csv2json(contents, {
      columns: true,
      bom: true,
      delimiter: options.delimiter || ',',
      // eslint-disable-next-line rulesdir/typography
      quote: '"',
      trim: true,
      relax_column_count: true,
    });
  } catch (err) {
    errors.push({
      message: 'Failed parsing: ' + err.message,
      internal: err.message,
    });
    return { errors, transactions: [] };
  }

  return { errors, transactions: data };
}

async function parseQIF(filepath) {
  let errors = [];
  let contents = await fs.readFile(filepath);

  let data;
  try {
    data = qif2json(contents);
  } catch (err) {
    errors.push({
      message: 'Failed parsing: doesnâ€™t look like a valid QIF file.',
      internal: err.stack,
    });
    return { errors, transactions: [] };
  }

  return {
    errors,
    transactions: data.transactions.map(trans => ({
      amount: trans.amount != null ? looselyParseAmount(trans.amount) : null,
      date: trans.date,
      payee_name: trans.payee,
      imported_payee: trans.payee,
      notes: trans.memo || null,
    })),
  };
}

async function parseOFX(filepath) {
  let { getOFXTransactions, initModule } = await import(
    /* webpackChunkName: 'xfo' */ 'node-libofx'
  );
  await initModule();

  let errors = [];
  let contents = await fs.readFile(filepath, 'binary');

  let data;
  try {
    data = getOFXTransactions(contents);
  } catch (err) {
    errors.push({
      message: 'Failed importing file',
      internal: err.stack,
    });
    return { errors };
  }

  return {
    errors,
    transactions: data.map(trans => ({
      amount: trans.amount,
      imported_id: trans.fi_id,
      date: trans.date ? dayFromDate(trans.date * 1000) : null,
      payee_name: trans.name,
      imported_payee: trans.name,
      notes: trans.memo || null,
    })),
  };
}
