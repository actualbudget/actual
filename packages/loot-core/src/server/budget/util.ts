import { safeNumber } from '../../shared/util';
import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return safeNumber(
    amounts.reduce((total, amount) => {
      return total + number(amount);
    }, 0),
  );
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}
