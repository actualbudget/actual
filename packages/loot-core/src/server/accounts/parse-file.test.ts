import * as d from 'date-fns';

import { amountToInteger } from '../../shared/util';
import * as db from '../db';
import * as prefs from '../prefs';

import { parseFile } from './parse-file';
import { reconcileTransactions } from './sync';

beforeEach(global.emptyDatabase());

// libofx spits out errors that contain the entire
// source code of the file in the stack which makes
// it hard to test.
let old = console.warn;
beforeAll(() => {
  console.warn = () => {};
});
afterAll(() => {
  console.warn = old;
});

async function getTransactions(accountId) {
  return db.runQuery(
    'SELECT * FROM transactions WHERE acct = ?',
    [accountId],
    true,
  );
}

async function importFileWithRealTime(
  accountId,
  filepath,
  dateFormat?: string,
) {
  // Emscripten requires a real Date.now!
  global.restoreDateNow();
  let { errors, transactions } = await parseFile(filepath, {
    enableExperimentalOfxParser: true,
  });
  global.restoreFakeDateNow();

  if (transactions) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    transactions = (transactions as any[]).map(trans => ({
      ...trans,
      amount: amountToInteger(trans.amount),
      date: dateFormat
        ? d.format(d.parse(trans.date, dateFormat, new Date()), 'yyyy-MM-dd')
        : trans.date,
    }));
  }

  if (errors.length > 0) {
    return { errors, added: [] };
  }

  let { added } = await reconcileTransactions(accountId, transactions);
  return { errors, added };
}

describe('File import', () => {
  test('qif import works', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });
    let { errors } = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/data.qif',
      'MM/dd/yy',
    );
    expect(errors.length).toBe(0);
    expect(await getTransactions('one')).toMatchSnapshot();
  });

  test('ofx import works', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });

    let { errors } = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/data.ofx',
    );
    expect(errors.length).toBe(0);
    expect(await getTransactions('one')).toMatchSnapshot();
  }, 45000);

  test('ofx import works (credit card)', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });

    let { errors } = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/credit-card.ofx',
    );
    expect(errors.length).toBe(0);
    expect(await getTransactions('one')).toMatchSnapshot();
  }, 45000);

  test('qfx import works', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });

    let { errors } = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/data.qfx',
    );
    expect(errors.length).toBe(0);
    expect(await getTransactions('one')).toMatchSnapshot();
  }, 45000);

  test('matches extensions correctly (case-insensitive, etc)', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });

    let res = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/best.data-ever$.QFX',
    );
    expect(res.errors.length).toBe(0);

    res = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/big.data.QiF',
      'MM/dd/yy',
    );
    expect(res.errors.length).toBe(0);

    res = await importFileWithRealTime('one', 'foo.txt');
    expect(res.errors.length).toBe(1);
    expect(res.errors[0].message).toBe('Invalid file type');
  }, 45000);

  test('handles non-ASCII characters', async () => {
    prefs.loadPrefs();
    await db.insertAccount({ id: 'one', name: 'one' });

    let { errors } = await importFileWithRealTime(
      'one',
      __dirname + '/../../mocks/files/8859-1.qfx',
      'yyyy-MM-dd',
    );
    expect(errors.length).toBe(0);
    expect(await getTransactions('one')).toMatchSnapshot();
  });
});
