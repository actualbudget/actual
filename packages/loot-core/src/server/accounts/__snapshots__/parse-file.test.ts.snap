// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File import handles non-ASCII characters 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": -2000,
    "category": null,
    "cleared": 1,
    "date": 20221019,
    "description": "id2",
    "error": null,
    "financial_id": "wSoKuCS77",
    "id": "id3",
    "imported_description": "Paiement facture/Carte prépayée",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": "PWW",
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
]
`;

exports[`File import ofx import works (credit card) 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": -600,
    "category": null,
    "cleared": 1,
    "date": 20230330,
    "description": "id2",
    "error": null,
    "financial_id": "57f91dca-0460-4692-be90-84165b7002",
    "id": "id3",
    "imported_description": "LOUISIANA STATE MUSEUM751 PLACE JOHN PAUL NEW ORLEANS 70116-3205LA USA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
]
`;

exports[`File import ofx import works 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": -3000,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id2",
    "error": null,
    "financial_id": "00092990122-30.00019012312798.01",
    "id": "id10",
    "imported_description": "PATIENT FIRST TOKEN 01/22 PURCHA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -377,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id3",
    "error": null,
    "financial_id": "00092990121-3.77019012312828.01",
    "id": "id11",
    "imported_description": "STARBUCKS STORE 07604 01/21 PURC",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -962,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id3",
    "error": null,
    "financial_id": "00092990121-9.62019012312831.78",
    "id": "id12",
    "imported_description": "STARBUCKS STORE 07604 01/21 PURC",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -11599,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id4",
    "error": null,
    "financial_id": "00090231800-115.99019012212841.40",
    "id": "id13",
    "imported_description": "VERIZON DES:PAYMENTREC ID:XXXXX3",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1034,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id5",
    "error": null,
    "financial_id": "00092990120-10.34019012212957.39",
    "id": "id14",
    "imported_description": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -2253,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id5",
    "error": null,
    "financial_id": "00092990120-22.53019012212967.73",
    "id": "id15",
    "imported_description": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -10871,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id6",
    "error": null,
    "financial_id": "00092990119-108.71019012212990.26",
    "id": "id16",
    "imported_description": "TMOBILE*AUTO PAY 01/19 PURCHASE",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -2500,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id7",
    "error": null,
    "financial_id": "00092990118-25.00019012213098.97",
    "id": "id17",
    "imported_description": "COUNTY WASTE 01/18 PURCHASE 804-",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -3238,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id8",
    "error": null,
    "financial_id": "00092990118-32.38019012213123.97",
    "id": "id18",
    "imported_description": "REGENCY MART CITGO 01/18 PURCHAS",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -607,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id9",
    "error": null,
    "financial_id": "00092990117-6.07019012213156.35",
    "id": "id19",
    "imported_description": "CHICK-FIL-A #01342 01/17 PURCHAS",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
]
`;

exports[`File import qfx import works 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": -3000,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id2",
    "error": null,
    "financial_id": "00092990122-30.00019012312798.01",
    "id": "id10",
    "imported_description": "PATIENT FIRST TOKEN 01/22 PURCHA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -377,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id3",
    "error": null,
    "financial_id": "00092990121-3.77019012312828.01",
    "id": "id11",
    "imported_description": "STARBUCKS STORE 07604 01/21 PURC",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -962,
    "category": null,
    "cleared": 1,
    "date": 20190123,
    "description": "id3",
    "error": null,
    "financial_id": "00092990121-9.62019012312831.78",
    "id": "id12",
    "imported_description": "STARBUCKS STORE 07604 01/21 PURC",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -11599,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id4",
    "error": null,
    "financial_id": "00090231800-115.99019012212841.40",
    "id": "id13",
    "imported_description": "VERIZON DES:PAYMENTREC ID:XXXXX3",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1034,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id5",
    "error": null,
    "financial_id": "00092990120-10.34019012212957.39",
    "id": "id14",
    "imported_description": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -2253,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id5",
    "error": null,
    "financial_id": "00092990120-22.53019012212967.73",
    "id": "id15",
    "imported_description": "URBAN FARMHOUSE NO 2 01/19 PURCH",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -10871,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id6",
    "error": null,
    "financial_id": "00092990119-108.71019012212990.26",
    "id": "id16",
    "imported_description": "TMOBILE*AUTO PAY 01/19 PURCHASE",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -2500,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id7",
    "error": null,
    "financial_id": "00092990118-25.00019012213098.97",
    "id": "id17",
    "imported_description": "COUNTY WASTE 01/18 PURCHASE 804-",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -3238,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id8",
    "error": null,
    "financial_id": "00092990118-32.38019012213123.97",
    "id": "id18",
    "imported_description": "REGENCY MART CITGO 01/18 PURCHAS",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -607,
    "category": null,
    "cleared": 1,
    "date": 20190122,
    "description": "id9",
    "error": null,
    "financial_id": "00092990117-6.07019012213156.35",
    "id": "id19",
    "imported_description": "CHICK-FIL-A #01342 01/17 PURCHAS",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
]
`;

exports[`File import qif import works 1`] = `
Array [
  Object {
    "acct": "one",
    "amount": -2500,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id2",
    "error": null,
    "financial_id": null,
    "id": "id19",
    "imported_description": "COUNTY WASTE 12/18 PURCHASE 804-8439288 VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -4500,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id3",
    "error": null,
    "financial_id": null,
    "id": "id20",
    "imported_description": "õSSELDORF KÖLN SALON - LIBBIE GR 12/18 PURCHASE RICHMOND VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -3133,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id4",
    "error": null,
    "financial_id": null,
    "id": "id21",
    "imported_description": "LOFT 0845 CHES 12/19 PURCHASE RICHMOND VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -631,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id5",
    "error": null,
    "financial_id": null,
    "id": "id22",
    "imported_description": "TARGET T- 1130 12/19 PURCHASE Richmond VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -4192,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id6",
    "error": null,
    "financial_id": null,
    "id": "id23",
    "imported_description": "TARGET DEBIT CRD ACH TRAN CO ID:XXXXX15170 POS TARGET -1018 RICHMOND V",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -2401,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id7",
    "error": null,
    "financial_id": null,
    "id": "id24",
    "imported_description": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9321 CO I",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1200,
    "category": null,
    "cleared": 1,
    "date": 20181219,
    "description": "id8",
    "error": null,
    "financial_id": null,
    "id": "id25",
    "imported_description": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 9703 CO I",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": 87485,
    "category": null,
    "cleared": 1,
    "date": 20181220,
    "description": "id9",
    "error": null,
    "financial_id": null,
    "id": "id26",
    "imported_description": "BKOFAMERICA MOBILE 12/20 3705340250 DEPOSIT *MOBILE VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1263,
    "category": null,
    "cleared": 1,
    "date": 20181220,
    "description": "id10",
    "error": null,
    "financial_id": null,
    "id": "id27",
    "imported_description": "AMAZON.COM*M21KD0HA0 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -3600,
    "category": null,
    "cleared": 1,
    "date": 20181220,
    "description": "id11",
    "error": null,
    "financial_id": null,
    "id": "id28",
    "imported_description": "TARGET DEBIT CRD DES:ACH TRAN ID:XXXXX0370723991 INDN:S LONG 6486 CO I",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -521,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id12",
    "error": null,
    "financial_id": null,
    "id": "id29",
    "imported_description": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -3470,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id13",
    "error": null,
    "financial_id": null,
    "id": "id30",
    "imported_description": "AMAZON.COM*M236431E2 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -541,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id12",
    "error": null,
    "financial_id": null,
    "id": "id31",
    "imported_description": "STARBUCKS STORE 07604 12/19 PURCHASE RICHMOND VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -10529,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id14",
    "error": null,
    "financial_id": null,
    "id": "id32",
    "imported_description": "AMAZON.COM*M208P98U1 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -736,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id15",
    "error": null,
    "financial_id": null,
    "id": "id33",
    "imported_description": "AMAZON.COM*M29QF7M71 AM 12/19 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1285,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id16",
    "error": null,
    "financial_id": null,
    "id": "id34",
    "imported_description": "AMAZON.COM*M20VB1851 AM 12/20 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -1299,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id17",
    "error": null,
    "financial_id": null,
    "id": "id35",
    "imported_description": "AMZN MKTP US*M27HH51C2 12/20 PURCHASE AMZN.COM/BILL WA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
  Object {
    "acct": "one",
    "amount": -6278,
    "category": null,
    "cleared": 1,
    "date": 20181221,
    "description": "id18",
    "error": null,
    "financial_id": null,
    "id": "id36",
    "imported_description": "CVS/PHARM 0154 12/21 PURCHASE RICHMOND VA",
    "isChild": 0,
    "isParent": 0,
    "location": null,
    "notes": null,
    "parent_id": null,
    "pending": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transferred_id": null,
    "type": null,
  },
]
`;
